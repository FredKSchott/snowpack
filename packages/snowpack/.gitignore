node_modules
pkg/*
!pkg/package.json
