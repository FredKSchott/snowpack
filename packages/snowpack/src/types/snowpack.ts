import type HttpProxy from 'http-proxy';
import {Plugin as RollupPlugin} from 'rollup';

export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Array<infer U>
    ? Array<DeepPartial<U>>
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : DeepPartial<T[P]>;
};

export type EnvVarReplacements = Record<string, string | number | true>;

export type SourceMappedCode = {code: string; map?: string};
export type SnowpackBuildMap = Record<string, string | SourceMappedCode>;

/** Standard file interface */
export interface SnowpackSourceFile {
  /** base extension (e.g. `.js`) */
  baseExt: string;
  /** file contents */
  contents: string;
  /** expanded extension (e.g. `.proxy.js` or `.module.css`) */
  expandedExt: string;
  /** if no location on disk, assume this exists in memory */
  locOnDisk: string;
}

export interface PluginLoadOptions {
  filePath: string;
  fileExt: string;
  isDev: boolean;
  isHmrEnabled: boolean;
}

export interface PluginTransformOptions {
  filePath: string;
  fileExt: string;
  contents: string | Buffer;
  isDev: boolean;
  isHmrEnabled: boolean;
}

export interface PluginRunOptions {
  isDev: boolean;
  isHmrEnabled: boolean;
}

/** map of extensions -> code (e.g. { ".js": "[code]", ".css": "[code]" }) */
export type PluginLoadResult = string | SnowpackBuildMap;

export interface PluginOptimizeOptions {
  buildDirectory: string;
}

export interface SnowpackPlugin {
  /** name of the plugin */
  name: string;
  /** Tell Snowpack how the load() function will resolve files. */
  resolve?: {
    /** file extensions that this load function takes as input (e.g. [".jsx", ".js", â€¦]) */
    input: string[];
    /** file extensions that this load function outputs (e.g. [".js", ".css"]) */
    output: string[];
  };
  /** load a file that matches resolve.input */
  load?(options: PluginLoadOptions): Promise<PluginLoadResult | null | undefined | void>;
  /** transform a file that matches resolve.input */
  transform?(options: PluginTransformOptions): Promise<string | null | undefined | void>;
  /** runs a command, unrelated to file building (e.g. TypeScript, ESLint) */
  run?(options: PluginRunOptions): Promise<unknown>;
  /** optimize the entire built application */
  optimize?(options: PluginOptimizeOptions): Promise<void>;
  /** Known dependencies that should be installed */
  knownEntrypoints?: string[];
}

export interface LegacySnowpackPlugin {
  defaultBuildScript: string;
  build?(options: PluginLoadOptions & {contents: string}): Promise<any>;
  bundle?(options: {
    srcDirectory: string;
    destDirectory: string;
    jsFilePaths: string[];
  }): Promise<any>;
}

/** Snowpack Build Plugin type */
export type SnowpackPluginFactory<PluginOptions = object> = (
  snowpackConfig: SnowpackConfig,
  pluginOptions?: PluginOptions,
) => SnowpackPlugin;

export type ProxyOptions = HttpProxy.ServerOptions & {
  // Custom on: {} event handlers
  on: Record<string, Function>;
};
export type Proxy = [string, ProxyOptions];

// interface this library uses internally
export interface SnowpackConfig {
  install: string[];
  extends?: string;
  exclude: string[];
  knownEntrypoints: string[];
  webDependencies?: {[packageName: string]: string};
  proxy: Proxy[];
  mount: Record<string, string>;
  alias: Record<string, string>;
  scripts: Record<string, string>;
  plugins: SnowpackPlugin[];
  devOptions: {
    secure: boolean;
    hostname: string;
    port: number;
    out: string;
    fallback: string;
    open: string;
    hmr: boolean;
  };
  installOptions: {
    dest: string;
    env: EnvVarReplacements;
    treeshake?: boolean;
    installTypes: boolean;
    sourceMap?: boolean | 'inline';
    externalPackage: string[];
    namedExports: string[];
    rollup: {
      plugins: RollupPlugin[]; // for simplicity, only Rollup plugins are supported for now
      dedupe?: string[];
    };
  };
  buildOptions: {
    baseUrl: string;
    webModulesUrl: string;
    clean: boolean;
    metaDir: string;
    minify: boolean;
    sourceMaps: boolean;
  };
  _extensionMap: Record<string, string>;
}

export interface CLIFlags extends Omit<Partial<SnowpackConfig['installOptions']>, 'env'> {
  help?: boolean; // display help text
  version?: boolean; // display Snowpack version
  reload?: boolean;
  config?: string; // manual path to config file
  env?: string[]; // env vars
  open?: string[];
  secure?: boolean;
}

export interface ImportMap {
  imports: {[packageName: string]: string};
}

export interface CommandOptions {
  cwd: string;
  config: SnowpackConfig;
  lockfile: ImportMap | null;
  pkgManifest: any;
}

/**
 * An install target represents information about a dependency to install.
 * The specifier is the key pointing to the dependency, either as a package
 * name or as an actual file path within node_modules. All other properties
 * are metadata about what is actually being imported.
 */
export type InstallTarget = {
  specifier: string;
  all: boolean;
  default: boolean;
  namespace: boolean;
  named: string[];
};

export type DependencyStats = {
  size: number;
  gzip: number;
  brotli?: number;
  delta?: number;
};

export type DependencyType = 'direct' | 'common';

export type DependencyStatsMap = {
  [filePath: string]: DependencyStats;
};

export type DependencyStatsOutput = Record<DependencyType, DependencyStatsMap>;
