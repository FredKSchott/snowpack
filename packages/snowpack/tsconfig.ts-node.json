{
  "ts-node": {
    "transpileOnly": true
  },
  "compilerOptions": {
    "target": "ES2015",
    "module": "CommonJS",
    "moduleResolution": "node",
    "esModuleInterop": true,
    "strict": true,
    "noImplicitAny": false,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "jsx": "react"
  },
  "include": ["src/**/*", "@types/**/*", "scripts/ts-node.bin.ts"],
  "exclude": ["node_modules"]
}
