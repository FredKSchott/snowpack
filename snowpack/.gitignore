lib/*
node_modules
vendor/generated/
vendor/types/
