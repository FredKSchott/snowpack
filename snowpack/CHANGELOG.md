# Changelog

> _For older releases, check our curated [release update thread](https://github.com/snowpackjs/snowpack/discussions/1183) or the raw [commit history](https://github.com/snowpackjs/snowpack/commits/main/snowpack)._

<!-- add changelog entries here - do not delete! -->

## snowpack@3.2.2 [2021-04-01]

* a643af33 - Detect Sass partial changes in dev (#3060) <Drew Powers>

## snowpack@3.2.1 [2021-03-31]

* 0663c1ac - fix HMR issue 

## snowpack@3.2.0 [2021-03-31]

* 56c5a231 - fix typo in changelog 
* c3971f10 - add snowpack release notes 
* 40f02cf4 - reorder changelog entries 
* a25f407d - Update CHANGELOG.md 
* c9dacc29 - Update CHANGELOG.md 
* a676d665 - fix website package.json 
* 68df7237 - support remote import if specifier has version in it 
* 16f58b99 - [ci] yarn format 
* d3859c8d - fix lint 
* 598f05a7 - Feat: import.meta.glob and import.meta.globEager support (#2881) <Nate Moore>
* 57a65bb8 - [ci] yarn format 
* 685f2200 - add docs, cleanup handler 
* 5fe83d4d - [ci] yarn format 
* 844cbcbe - routes support upgrade event (#2988) <dishuostec>
* a17dd97c - Fix: Only import directory/index.js as a fallback mechanism (#3018) <FranÃ§ois Wouts>
* e340dea9 - Do not exclude folders starting with a dot (#2962) <Matthew Phillips>
* fc6c1417 - Improve CSS error message (#2973) <Drew Powers>
* b41b14b0 - Await promises in local source (#2972) <Drew Powers>

## snowpack@3.1.2 [2021-03-20]

* e50cb745 - fix better handling for unscannable cjs packages
* 1232e252 - add back cjs-cjs compat from an earlier pr (#2934)

## snowpack@3.1.1 [2021-03-19]

* f53498ba - Fix preact regression (#2926) 

## snowpack@3.1.0 [2021-03-19]

* 872bfc74 - [ci] yarn format <natemoo-re>
* c8aee80f - Support compound output file extensions (#2593) <Edward Faulkner>
* 8dbd6af3 - [ci] yarn format <natemoo-re>
* 88d3a9a9 - handle legacy package urls (#2915) 
* 094b7a12 - Add `openUrl` option to config (#2902) <Dan Marshall>
* aba028c4 - [ci] yarn format <natemoo-re>
* 4de244ef - Add support for custom TLS key files via devOptions.secure = { cert, key } (#2356) <Aral Balkan>
* da049ce4 - [ci] yarn format <natemoo-re>
* b569d35b - add getUrlForPackage utility (#2913) <Nate Moore>
* d826a60c - Remove enumerability of converted ESM in SSR (#2920) <Matthew Phillips>
* 941110cd - fix esbuild 0.9.x breaking change (#2914) 
* 35289b4b - improve import scanner perf (#2900) 
* a800bf3d - finalize picomatch support (#2912) 
* d33076a3 - Speed up exclusions with picomatch  (#2904) <Ismail Syed>
* 84e034cc - [ci] yarn format 
* f35bc421 - fix debug log 
* 81fd028e - [ci] yarn format <natemoo-re>
* 7c8236e8 - Add config.env property to support non-prefixed env variables (#2390) <moonrailgun>
* 337cfd7a - only target actual imports (Regex Fix) (#2817) <Danzo7>
* aa953cca - refactor: update to esbuild 0.9 (#2886) <Ludovico Fischer>
* cd4f344b - [ci] yarn format <natemoo-re>
* 0fce7efd - Support buildOptions.jsxInject in config (#2884) <Nate Moore>
* dc503c84 - speed up mount scanning with fdir (#2876) 
* e8dead62 - improve handling of "." & ".." imports (#2877) 
* a178f286 - run transform on all dependencies (#2878) 
* 27321b83 - collect known imports by package and version 
* 3a6ad53a - update (#2363) <Myou Aki>
* b2738967 - improve cjs<>esm conversion of named exports (#2859) 
* 3ba7fede - feat: support specifying sourcemap type in config.optimize.sourcemap (#2793) <yqrashawn>
* 03ceb62d - Resolve missing extensions (#2839) <Drew Powers>
* 28689888 - [ci] yarn format 
* 689f58e9 - SSR fixes (#2853) <Matthew Phillips>
* 3549e43c - Fix Mac OS devOptions.open bug (#2798) <joakim-mjardner>
* 21461ce5 - Update @snowpack/plugin-postcss API (#2854) <Drew Powers>
* 9fbe93fb - fix: revert snowpack to v3.0.13 for CI tests (#2851) <Nate Moore>
* 339d3cf1 - [ci] yarn format <natemoo-re>
* ea4ebd17 - Proper loading of ESM config files (#2834) <Nate Moore>
* 8f022949 - snowpack@3.1.0-pre.12 
* 9d987e4a - fix watching logic and dashboard output 
* 15428ae4 - [ci] yarn format 
* bea1c56c - Simplify. cleanup, enhance snowpack internals (#2707) 
* 43677e89 - Add import support for .interface files, alongside .svelte, and .vue. (#2380) <Aral Balkan>
* 3d457a8e - fix(ssr-loader): support css modules (#2528) <Nate Moore>
* 23d7b450 - fix: recognise dot as a valid relative import (#2662) <Jon Rimmer>
* 582808c3 - [ci] yarn format 
* deabcb0e - don't escape utf-8 characters by cheerio and esbuild (#2755) <bsl-zcs>
* d35d50b4 - [ci] yarn format 
* b756b49d - Fix snowpack add/rm for npm packages with @ prefix (#2665) <Nigel>
* 58df86d4 - add a tip if someone uses process.env 
* fdda447d - dont match meta paths to routes 

## snowpack@3.0.13 [2021-02-23]

* 74d0be10 - add resolve for bundled package 


## snowpack@3.0.12 [2021-02-23]

* d3b6f769 - Support packages that use export maps but have no main (#2659) <Matthew Phillips>
* 2ae933be - add true logger error message 
* 0de849ee - Catch WebSocket message parse errors (#2470) <Ben Foxall>
* da75ff23 - remove circular dependencies in util (#2366) 
* bfa32aff - proper default browser support for mac (#2364) 
* 1c44c22d - fix bad url access in import maps, snowpack add 
* 76dca31a - add octet-stream handling 
* 10f3461b - [ci] yarn format 
* 615f13a7 - Add isHmrEnabled and isSSR to plugin transform. (#2332) <Chris Thomas>
* 65008c63 - fix: prettier snowpack/src/build dir (#2346) <ZYSzys>
* 690d3b1d - update prettier to use explicit formatting 
* f07105a9 - Add root level export of clearCache (#2330) <Kevin Scott>
* 137139b4 - [ci] yarn format 
* e0f01181 - fix: prettier format (#2316) <ZYSzys>
* 0fc858cf - fix bad optimize external filter 
* fb167098 - load plugins relative to config root 
* 85ddf618 - add JS API docs (#2315) 
* 49b29f97 - [ci] yarn format 

## snowpack@3.0.11 [2021-01-16]

- 58a36d4c - fix bad build conditional check
- 37f7a55d - remove empty build folders from build
- 6e01eeb5 - improve build watch output
- 137c58b0 - fix init file issue
- e2b74e7b - add fallback deprecation
- 968720cf - remove rollup from bundle to pass peerDep checks
- 70d25dad - Resolves #2265 (#2289) <Josh Wilson>
- 7b38b486 - get tests passing
- e6b618c5 - [ci] yarn format

## snowpack@3.0.10 [2021-01-13]

- 1d071628 - cleanup the bundled build before publishing
- beb3f9a1 - stop ignoring postcss in generated bundle
- 2b797521 - [ci] yarn format

## snowpack@3.0.9 [2021-01-13]

- d26c7733 - [ci] yarn format

## snowpack@3.0.8 [2021-01-13]

- c2124bd7 - json fix
- e040f071 - Fix postcss-modules being required as a dependency in projects (#2257) <David Bailey>
- c566aa10 - add bundled type support for esinstall and skypack, remove as full deps
- ecf85d52 - move open back to regular dependency
- b0f7280c - [ci] yarn format

## snowpack@3.0.7 [2021-01-13]

- 61bd4c3b - [ci] yarn format
- eb5b366a - cleanup types, release script
- 20e3ed21 - move skypack and esbuild into real dependencies for now
- e63d70fd - fix direct foo.svelte.css references
- 01a80448 - [ci] yarn format

## snowpack@3.0.6 [2021-01-13]

- 8f614490 - final cleanup before v3 goes out
- 15d77ea3 - cleanup
