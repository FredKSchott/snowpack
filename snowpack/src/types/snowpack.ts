import type HttpProxy from 'http-proxy';
import type * as http from 'http';
import type {InstallOptions} from 'esinstall';
import type {RawSourceMap} from 'source-map';

export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Array<infer U>
    ? Array<DeepPartial<U>>
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : DeepPartial<T[P]>;
};

export type SnowpackBuiltFile = {
  code: string | Buffer;
  map?: string;
};

export type SnowpackBuildMap = Record<string, SnowpackBuiltFile>;

/** Standard file interface */
export interface SnowpackSourceFile<Type = string | Buffer> {
  /** base extension (e.g. `.js`) */
  baseExt: string;
  /** file contents */
  contents: Type;
  /** expanded extension (e.g. `.proxy.js` or `.module.css`) */
  expandedExt: string;
  /** if no location on disk, assume this exists in memory */
  locOnDisk: string;
}

export interface PluginLoadOptions {
  filePath: string;
  fileExt: string;
  isDev: boolean;
  isSSR: boolean;
  isHmrEnabled: boolean;
}

export interface PluginTransformOptions {
  id: string;
  fileExt: string;
  contents: string | Buffer;
  isDev: boolean;
  isHmrEnabled: boolean;
}

export interface PluginRunOptions {
  isDev: boolean;
  isHmrEnabled: boolean;
}

/** map of extensions -> code (e.g. { ".js": "[code]", ".css": "[code]" }) */
export type PluginLoadResult = SnowpackBuildMap;

export type PluginTransformResult = {contents: string; map: string | RawSourceMap};

export interface PluginOptimizeOptions {
  buildDirectory: string;
}

export interface SnowpackPlugin {
  /** name of the plugin */
  name: string;
  /** Tell Snowpack how the load() function will resolve files. */
  resolve?: {
    /**
       file extensions that this load function takes as input (e.g. [".jsx",
       ".js", â€¦])
     */
    input: string[];
    /**
       file extensions that this load function outputs (e.g. [".js", ".css"])
     */
    output: string[];
  };
  /** load a file that matches resolve.input */
  load?(options: PluginLoadOptions): Promise<PluginLoadResult | string | null | undefined | void>;
  /** transform a file that matches resolve.input */
  transform?(
    options: PluginTransformOptions,
  ): Promise<PluginTransformResult | string | null | undefined | void>;
  /** runs a command, unrelated to file building (e.g. TypeScript, ESLint) */
  run?(options: PluginRunOptions): Promise<unknown>;
  /** optimize the entire built application */
  optimize?(options: PluginOptimizeOptions): Promise<void>;
  /** cleanup any long-running instances/services before exiting.  */
  cleanup?(): void | Promise<void>;
  /** Known dependencies that should be installed */
  knownEntrypoints?: string[];
  /** read and modify the Snowpack config object */
  config?(snowpackConfig: SnowpackConfig): void;
}

export interface LegacySnowpackPlugin {
  defaultBuildScript: string;
  build?(options: PluginLoadOptions & {contents: string | Buffer}): Promise<any>;
  bundle?(options: {
    srcDirectory: string;
    destDirectory: string;
    jsFilePaths: string[];
  }): Promise<any>;
}

/** Snowpack Build Plugin type */
export type SnowpackPluginFactory<PluginOptions = object> = (
  snowpackConfig: SnowpackConfig,
  pluginOptions?: PluginOptions,
) => SnowpackPlugin;

export type ProxyOptions = HttpProxy.ServerOptions & {
  // Custom on: {} event handlers
  on: Record<string, Function>;
};
export type Proxy = [string, ProxyOptions];

// interface this library uses internally
export interface SnowpackConfig {
  install: string[];
  extends?: string;
  exclude: string[];
  knownEntrypoints: string[];
  webDependencies?: {[packageName: string]: string};
  proxy: Proxy[];
  mount: Record<string, string>;
  alias: Record<string, string>;
  scripts: Record<string, string>;
  plugins: SnowpackPlugin[];
  devOptions: {
    secure: boolean;
    hostname: string;
    port: number;
    out: string;
    fallback: string;
    open: string;
    output: 'stream' | 'dashboard';
    hmr?: boolean;
    hmrDelay: number;
    hmrPort: number;
    hmrErrorOverlay: boolean;
  };
  installOptions: InstallOptions;
  buildOptions: {
    baseUrl: string;
    webModulesUrl: string;
    clean: boolean;
    metaDir: string;
    minify: boolean;
    sourceMaps: boolean;
    watch: boolean;
  };
  /** EXPERIMENTAL - This section is experimental and not yet finalized. May change across minor versions. */
  experiments: {
    /** (EXPERIMENTAL) If true, "snowpack build" should build your site for SSR. */
    ssr: boolean;
    /** (EXPERIMENTAL) Custom request handler for the dev server. */
    app?: (
      req: http.IncomingMessage,
      res: http.ServerResponse,
      next: (err?: Error) => void,
    ) => unknown;
  };
  _extensionMap: Record<string, string>;
}

export interface CLIFlags extends Omit<InstallOptions, 'env'> {
  help?: boolean; // display help text
  version?: boolean; // display Snowpack version
  reload?: boolean;
  config?: string; // manual path to config file
  env?: string[]; // env vars
  open?: string[];
  secure?: boolean;
  verbose?: boolean;
  quiet?: boolean;
}

export interface ImportMap {
  imports: {[packageName: string]: string};
}

export interface CommandOptions {
  cwd: string;
  config: SnowpackConfig;
  lockfile: ImportMap | null;
  pkgManifest: any;
}

export type LoggerLevel = 'debug' | 'info' | 'warn' | 'error' | 'silent'; // same as Pino
export type LoggerEvent = 'debug' | 'info' | 'warn' | 'error';
export interface LoggerOptions {
  /** (optional) change name at beginning of line */
  name?: string;
}
