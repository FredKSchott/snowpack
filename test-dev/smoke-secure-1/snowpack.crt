-----BEGIN CERTIFICATE-----
MIIERzCCAq+gAwIBAgIQUSVq6njLtPS7s+9BIHjGBzANBgkqhkiG9w0BAQsFADB1
MR4wHAYDVQQKExVta2NlcnQgZGV2ZWxvcG1lbnQgQ0ExJTAjBgNVBAsMHGFyYWxA
ZGV2LmFyLmFsIChBcmFsIEJhbGthbikxLDAqBgNVBAMMI21rY2VydCBhcmFsQGRl
di5hci5hbCAoQXJhbCBCYWxrYW4pMB4XDTIwMTIyMTE0MTcxMloXDTIzMDMyMTE0
MTcxMlowUDEnMCUGA1UEChMebWtjZXJ0IGRldmVsb3BtZW50IGNlcnRpZmljYXRl
MSUwIwYDVQQLDBxhcmFsQGRldi5hci5hbCAoQXJhbCBCYWxrYW4pMIIBIjANBgkq
hkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAzw5c6d2sxz7jWayEzyuZgod5bTW0VqQ8
I5WMuflQs/oiEX8OK9IPKkJ0lzDYe70jEbUlsnIPGP2XgyEwja1GLGqyt0BqdVsH
8vhmtBD0pICV5jA+R5v5o4jU07wsgHMC9rR4M5DPag3RjJWYsx7l5aosyEgMIpoF
t1r4XUukl2qFRlM8xXqywFyy0IfnT187ijh7VHT7Ibm+HHcHkFqbF9oqVUkR+NsB
0SEo4CEio5ufMCG0gmaqe4A81YGuLdUYYzPF5KOF4Hh1CAjXhKJr8dyETvixjZO3
40DmsguPbr5+xedof6USr0zuel56AA2Z3mwT0C8zndTd4AhpWx9GzQIDAQABo3gw
djAOBgNVHQ8BAf8EBAMCBaAwEwYDVR0lBAwwCgYIKwYBBQUHAwEwDAYDVR0TAQH/
BAIwADAfBgNVHSMEGDAWgBT2Ac1DO7MgFVDNm/l0GXsYfedtDzAgBgNVHREEGTAX
gglsb2NhbGhvc3SHBH8AAAGHBMCoAvkwDQYJKoZIhvcNAQELBQADggGBAEy0WzWB
ylaXs/VVFU5bXHCoVeSfMZlR0UOdIZ5CaAnzYx8YYVTKzR40hdI8w360BZkcPhcf
eESsx4lnHFgndvlIsrNRx5BxBVVRk6dzzgQMsQ3gr6auA+gCkzxNjXjB9eGwjVhU
rIyTQVDFhtAVq4rdBJd6dIipL8rlUuCXluVhBCcRfi3jiSPV2wNFvqGpPHN6X7yR
8j56zP7reSeL5CczdGbgJNLzp+jimz9NYLCuY97hURTTGeNV5+QGbgz18Uojjall
rgbWyrxIvf4SRxVOH2Y2htYtuinhB1UHQH8IuXpnBzoT1wAzT2tAxSy+Yi+QT5EK
kG0zVqqtYHkkHkBwnnGI0IA3hvQtvTsBCWZcQHM+1zK4BaVJJvtV60/Ejy8R1phw
2NhexVMUcLIrNNVZxS+O8l+Z5v0rBuL+rzuEB90idKWHCinJ0b4D9y7vUXdcXV6H
x4cunqRnZTYoHHZEY3YwKfHF5ICsToEKLDDgcmfJQZBEIepcmeyOtA/9sg==
-----END CERTIFICATE-----
