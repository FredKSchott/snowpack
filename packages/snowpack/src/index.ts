import * as colors from 'kleur/colors';
import path from 'path';
import yargs from 'yargs-parser';
import {addCommand, rmCommand} from './commands/add-rm';
import {command as buildCommand} from './commands/build';
import {command as devCommand} from './commands/dev';
import {command as installCommand} from './commands/install';
import {loadAndValidateConfig} from './config.js';
import {logger} from './logger';
import {CLIFlags} from './types/snowpack';
import {clearCache, readLockfile} from './util.js';

export {install as unstable_installCommand} from './commands/install';
export {createConfiguration} from './config.js';
export * from './types/snowpack';

const cwd = process.cwd();

function printHelp() {
  logger.info(
    `
${colors.bold(`snowpack`)} - A faster build system for the modern web.

  Snowpack is best configured via config file.
  But, most configuration can also be passed via CLI flags.
  ðŸ“– ${colors.dim('https://www.snowpack.dev/#configuration')}

${colors.bold('Commands:')}
  snowpack dev          Develop your app locally.
  snowpack build        Build your app for production.
  snowpack install      (Advanced) Install web-ready dependencies.

${colors.bold('Flags:')}
  --config [path]       Set the location of your project config file.
  --help                Show this help message.
  --version             Show the current version.
  --reload              Clear Snowpack's local cache (troubleshooting).
  --verbose             View debug info (where available)
  --quiet               Donâ€™t output anything (dev server will still log minimally)
    `.trim(),
  );
}

export async function cli(args: string[]) {
  // parse CLI flags
  const cliFlags = yargs(args, {
    array: ['install', 'env', 'exclude', 'externalPackage'],
  }) as CLIFlags;
  if (cliFlags.help) {
    printHelp();
    process.exit(0);
  }
  if (cliFlags.version) {
    logger.info(require('../package.json').version);
    process.exit(0);
  }
  if (cliFlags.reload) {
    logger.info(colors.yellow('! clearing cache...'));
    await clearCache();
  }

  // Load the current package manifest
  let pkgManifest: any;
  try {
    pkgManifest = require(path.join(cwd, 'package.json'));
  } catch (err) {
    logger.error(`package.json not found in directory: ${cwd}. Run \`npm init -y\` to create one.`);
    process.exit(1);
  }

  const cmd = cliFlags['_'][2];

  // Set this early -- before config loading -- so that plugins see it.
  if (cmd === 'build') {
    process.env.NODE_ENV = process.env.NODE_ENV || 'production';
  }
  if (cmd === 'dev') {
    process.env.NODE_ENV = process.env.NODE_ENV || 'development';
  }

  const commandOptions = {
    cwd,
    config: loadAndValidateConfig(cliFlags, pkgManifest),
    lockfile: await readLockfile(cwd),
    pkgManifest,
    logger,
  };

  if (cliFlags.verbose) logger.level = 'debug';
  if (cliFlags.quiet) logger.level = 'silent';

  if (cmd === 'add') {
    await addCommand(cliFlags['_'][3], commandOptions);
    return;
  }
  if (cmd === 'rm') {
    await rmCommand(cliFlags['_'][3], commandOptions);
    return;
  }

  if (cliFlags['_'].length > 3) {
    logger.error(`Unexpected multiple commands`);
    process.exit(1);
  }

  if (cmd === 'build') {
    await buildCommand(commandOptions);
    return;
  }
  if (cmd === 'dev') {
    await devCommand(commandOptions);
    return;
  }
  if (cmd === 'install' || !cmd) {
    await installCommand(commandOptions);
    return;
  }

  logger.error(`Unrecognized command: ${cmd}`);
  process.exit(1);
}
