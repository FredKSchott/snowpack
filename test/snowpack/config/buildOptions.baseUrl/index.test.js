const {testFixture} = require('../../../fixture-utils');

describe('buildOptions.baseUrl', () => {
  beforeAll(() => {
    // Needed until we make Snowpack's JS Build Interface quiet by default
    require('snowpack').logger.level = 'warn';
  });

  describe.each([
    ['local URL path', '/static/'],
    ['remote URL', 'https://www.example.com'],
  ])('%s', (_, baseUrl) => {
    const config = {buildOptions: {baseUrl}};

    it('generates correct relative package imports', async () => {
      const result = await testFixture(config, `import {flatten} from 'array-flatten';`);
      expect(result['index.js']).toMatchSnapshot();
    });

    it('generates correct <link> attributes', async () => {
      const result = await testFixture(config, {
        'index.html': `
<!DOCTYPE html>
<html lang="en">
  <head>
    <link rel="icon" href="%PUBLIC_URL%/favicon.ico" />
    <link rel="stylesheet" type="text/css" href="%PUBLIC_URL%/index.css" />
  </head>
</html>`,
      });
      expect(result['index.html']).toMatchSnapshot();
    });

    it('generates correct <script> attributes', async () => {
      const result = await testFixture(config, {
        'index.html': `
<!DOCTYPE html>
<html lang="en">
  <body><script type="module" src="%PUBLIC_URL%/_dist_/index.js"></script></body>
</html>`,
      });
      expect(result['index.html']).toMatchSnapshot();
    });

    it('generates correct import proxies', async () => {
      const result = await testFixture(config, {
        'index.js': `import logo from './logo.png';`,
        'logo.png': Buffer.from(''),
      });

      expect(result['index.js']).toMatchSnapshot();
      expect(result['logo.png.proxy.js']).toMatchSnapshot();
    });

    it('generates correct import.meta.env imports', async () => {
      const result = await testFixture(config, `console.log(import.meta.env)`);
      expect(result['index.js']).toMatchSnapshot();
    });
  });
});
