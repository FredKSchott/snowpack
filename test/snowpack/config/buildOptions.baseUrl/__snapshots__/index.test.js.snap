// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildOptions.baseUrl local URL path generates correct <link> attributes 1`] = `
"
<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <link rel=\\"icon\\" href=\\"/static/favicon.ico\\" />
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/static/index.css\\" />
  </head>
</html>"
`;

exports[`buildOptions.baseUrl local URL path generates correct <script> attributes 1`] = `
"
<!DOCTYPE html>
<html lang=\\"en\\">
  <body><script type=\\"module\\" src=\\"/static/_dist_/index.js\\"></script></body>
</html>"
`;

exports[`buildOptions.baseUrl local URL path generates correct import proxies 1`] = `"import logo from './logo.png.proxy.js';"`;

exports[`buildOptions.baseUrl local URL path generates correct import proxies 2`] = `"export default \\"/static/logo.png\\";"`;

exports[`buildOptions.baseUrl local URL path generates correct import.meta.env imports 1`] = `
"import * as __SNOWPACK_ENV__ from './_snowpack/env.js';

console.log(__SNOWPACK_ENV__)"
`;

exports[`buildOptions.baseUrl local URL path generates correct relative package imports 1`] = `"import {flatten} from './_snowpack/pkg/array-flatten.js';"`;

exports[`buildOptions.baseUrl remote URL generates correct <link> attributes 1`] = `
"
<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <link rel=\\"icon\\" href=\\"https://www.example.com/favicon.ico\\" />
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"https://www.example.com/index.css\\" />
  </head>
</html>"
`;

exports[`buildOptions.baseUrl remote URL generates correct <script> attributes 1`] = `
"
<!DOCTYPE html>
<html lang=\\"en\\">
  <body><script type=\\"module\\" src=\\"https://www.example.com/_dist_/index.js\\"></script></body>
</html>"
`;

exports[`buildOptions.baseUrl remote URL generates correct import proxies 1`] = `"import logo from './logo.png.proxy.js';"`;

exports[`buildOptions.baseUrl remote URL generates correct import proxies 2`] = `"export default \\"https://www.example.com/logo.png\\";"`;

exports[`buildOptions.baseUrl remote URL generates correct import.meta.env imports 1`] = `
"import * as __SNOWPACK_ENV__ from './_snowpack/env.js';

console.log(__SNOWPACK_ENV__)"
`;

exports[`buildOptions.baseUrl remote URL generates correct relative package imports 1`] = `"import {flatten} from './_snowpack/pkg/array-flatten.js';"`;
